src/mlpack/tests/imputation_test.cpp

```diff
@@ -0,0 +1,242 @@
+
+  arma::mat input;
+  arma::mat output;
+  string missingValue = "a";
+  double customValue = 99;
+  size_t feature = 0;
+
+  DatasetInfo info;
+  BOOST_REQUIRE(data::Load("test_file.csv", input, info) == true);
+
+  BOOST_REQUIRE_EQUAL(input.n_rows, 3);
+  BOOST_REQUIRE_EQUAL(input.n_cols, 3);
+
+  /* TODO: Connect Load with the new DatasetMapper instead of DatasetInfo*/
+
+  Imputer<double,
+          DatasetInfo,
+          CustomImputation<double>> impu(info);
+  impu.Impute(input, output, missingValue, customValue, feature);
+  // Remove the file.
+  remove("test_file.csv");
+}
+
+/**
+ * Make sure a CSV is loaded correctly.
+ */
+BOOST_AUTO_TEST_CASE(CustomImputationTest)
+{
+  arma::mat input("3.0 0.0 2.0 0.0;"
+                  "5.0 6.0 0.0 6.0;"
+                  "9.0 8.0 4.0 8.0;");
+  arma::mat outputT; // assume input is transposed
+  arma::mat output;  // assume input is not transposed
+  double customValue = 99;
+  double mappedValue = 0.0;
+
+  CustomImputation<double> imputer;
+
+  // transposed
+  imputer.Apply(input, outputT, mappedValue, customValue, 0/*dimension*/, true);
+
+  BOOST_REQUIRE_CLOSE(outputT(0, 0), 3.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(0, 1), 99.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(0, 2), 2.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(0, 3), 99.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(1, 0), 5.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(1, 1), 6.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(1, 2), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(1, 3), 6.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(2, 0), 9.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(2, 1), 8.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(2, 2), 4.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(2, 3), 8.0, 1e-5);
+
+  // not transposed
+  imputer.Apply(input, output, mappedValue, customValue, 1/*dimension*/, false);
+
+  BOOST_REQUIRE_CLOSE(output(0, 0), 3.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(0, 1), 99.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(0, 2), 2.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(0, 3), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 0), 5.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 1), 6.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 2), 0.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(1, 3), 6.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(2, 0), 9.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(2, 1), 8.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(2, 2), 4.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(output(2, 3), 8.0, 1e-5);
+}
+
+/**
+ * Make sure a CSV is loaded correctly.
+ */
+BOOST_AUTO_TEST_CASE(MeanImputationTest)
+{
+  arma::mat input("3.0 0.0 2.0 0.0;"
+                  "5.0 6.0 0.0 6.0;"
+                  "9.0 8.0 4.0 8.0;");
+  arma::mat outputT; // assume input is transposed
+  arma::mat output;  // assume input is not transposed
+  double mappedValue = 0.0;
+
+  MeanImputation<double> imputer;
+
+  // transposed
+  imputer.Apply(input, outputT, mappedValue, 0/*dimension*/, true);
+
+  BOOST_REQUIRE_CLOSE(outputT(0, 0), 3.0, 1e-5);
+  BOOST_REQUIRE_CLOSE(outputT(0, 1), 2.5, 1e-5);
```

Comment:
No matter how I calculate, the mean value should be 5.66666667
3+5+9+6+8+2+4+6+8 = 51
51/9 =  5.66666667

Maybe it is because to you forgot to initialize sum?